File: Week02/BOJ_오르막 수/won.java

import java.io.*;

public class Main {
    static int[][] arr;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 한 줄을 읽어와서 공백을 기준으로 분리
        String[] input = br.readLine().split(" ");

        // 각각을 N, M으로 변환
        int N = Integer.parseInt(input[0]);
        arr = new int[N+1][10]; // 행의 idx는 1부터 사용함

        // N =1 일때 작업(전부 1로 하기)
        for (int i = 0; i <= 9; i++) {
            arr[1][i] = 1;
        }

        // N >= 2 일 경우
        for(int k = 2; k <= N; k++){
            for (int i = 0; i <= 9; i++) {
                int calSumResult =cal_sum(k - 1, i ) ;
                // 각 단계 에서도 MOD 연산을 하는 것이 좋다. 왜냐면 안하면 arr 안의 값이 너무 커짐
                arr[k][i] = calSumResult % 10007 ;
            }
        }

        int ans = cal_sum(N,9) % 10007;

        System.out.println(ans);
    }

    // cal_sum: k번째 행에서 0~n번째 값의 합 (arr[k][0] ~ arr[k][n])
    private static int cal_sum(int k, int n) {
        int sum = 0;
        for (int i = 0; i <= n; i++) {
            sum+=arr[k][i];
        }

        return sum;
    }
}

File: Week03/BOJ_가장 긴 감소하는 부분 수열/won.py
'''
가장 긴 감소하는 부분 수열
dp 배열을 생성한다.
numbers에서 1 ~ 끝번 까지의 인덱스 탐색
idx = 5이면 numbers[5]와 numbers[0] ~ numbers[4]을 비교한다.
numbers[5] 보다 큰 값이 존재하고 그 인덱스가 3이면
dp[5] = max(dp[5], dp[3]+1) 이런식으로 dp 배열을 완성해나감.
최종 답: dp 배열에서 가장 큰 값
'''

# 첫 번째 입력: 수열의 크기
N = int(input())

# 두 번째 입력: 수열을 이루고 있는 것
numbers = list(map(int, input().split()))

dp = [1] * N

for i in range(1, N):
    for k in range(0,i):
        if(numbers[k] > numbers[i]):
            dp[i] = max(dp[i], dp[k]+1)



# 결과 출력
print(max(dp))
